xcell/cls/cl.py

```diff
@@ -323,28 +324,71 @@ def _compute_workspace(self, spin0=False, read_unbinned_MCM=True):
         else:
             fname = os.path.join(self.outdir, f'w__{mask1}__{mask2}.fits')
         w = nmt.NmtWorkspace()
-        if self.recompute_mcm or (not os.path.isfile(fname)):
-            n_iter = self.data.data['sphere']['n_iter_mcm']
-            l_toeplitz, l_exact, dl_band = self.data.check_toeplitz('cls')
-            if spin0:
-                m1, m2 = self.get_mappers()
-                msk1 = m1.get_mask()
-                msk2 = m2.get_mask()
-                f1 = nmt.NmtField(msk1, None, spin=0)
-                f2 = nmt.NmtField(msk2, None, spin=0)
-            else:
-                f1, f2 = self.get_nmt_fields()
-            w.compute_coupling_matrix(f1, f2, self.b, n_iter=n_iter,
-                                      l_toeplitz=l_toeplitz, l_exact=l_exact,
-                                      dl_band=dl_band)
-            # Recheck again in case other process has started writing it
-            if (not os.path.isfile(fname)):
-                w.write_to(fname)
-            self.recompute_mcmc = False
+        if (not self.recompute_mcm) and os.path.isfile(fname):
+            tools.read_wsp(w, fname, read_unbinned_MCM=read_unbinned_MCM)
+            if w.wsp is not None:
+                return w
+
+        n_iter = self.data.data['sphere']['n_iter_mcm']
+        l_toeplitz, l_exact, dl_band = self.data.check_toeplitz('cls')
+        if spin0:
+            m1, m2 = self.get_mappers()
+            msk1 = m1.get_mask()
+            msk2 = m2.get_mask()
+            f1 = nmt.NmtField(msk1, None, spin=0)
+            f2 = nmt.NmtField(msk2, None, spin=0)
```

Comment:
is this simply a reordering of the above, or was there a bug?

Reply:
It's to avoid crashing the run if the workspace in the disk is corrupted. If after reading the workspace it is corrupted, it will remove the file and compute the workspace again. Before it would have tried to use the corrupted loaded workspace (and fail)